### PR TITLE
Cloud iot core

### PR DESCRIPTION
**Pull request by gguuss**

### CODE REVIEW

**Comment by gguuss** [comment]
Requesting review, this is the API-client based sample I mentioned.

File: iot/api-client/README.md [hunk 1]
```diff
@@ -0,0 +1,37 @@
+# Cloud IoT Core Python Samples
```
Comment (jonparrott) [inline iot/api-client/README.md]: Why not use autogenned readmes?

File: iot/api-client/generate_keys.sh [hunk 2]
```diff
@@ -0,0 +1,20 @@
+#!/bin/bash
```
Comment (jonparrott) [inline iot/api-client/generate_keys.sh]: For portability (windows) I'm wondering if it's easy to do this in Windows? No action needed, but just curious.

File: iot/api-client/manager/cloudiot_device_manager_example.py [hunk 3]
```diff
@@ -0,0 +1,330 @@
...
+    from Google Cloud Platform's API Manager's Credentials page.
+"""
+
+import argparse
+import sys
+import time
+
+from googleapiclient import discovery
+from googleapiclient.errors import HttpError
+from oauth2client.service_account import ServiceAccountCredentials
```
Comment (jonparrott) [inline iot/api-client/manager/cloudiot_device_manager_example.py]: oauth2client is verboten.

File: iot/api-client/manager/cloudiot_device_manager_example.py [hunk 4]
```diff
@@ -0,0 +1,330 @@
...
+    """Administer a set of devices for a device registry."""
+
+    def __init__(
+            self, project_id, registry_id, cloud_region,
+            service_account_json, api_key, pubsub_topic):
+        """Lookup or create a device registry for the given project."""
+        self.parent = 'projects/{}/locations/{}'.format(
+                project_id, cloud_region)
+        self.full_name = '{}/registries/{}'.format(self.parent, registry_id)
+        credentials = ServiceAccountCredentials.from_json_keyfile_name(
```
Comment (jonparrott) [inline iot/api-client/manager/cloudiot_device_manager_example.py]: use `google.oauth2.service_account.Credentials.from_service_account_file`.

File: iot/api-client/manager/cloudiot_device_manager_example.py [hunk 5]
```diff
@@ -0,0 +1,330 @@
...
+                    'Could not load service account credential from {}'
+                    .format(service_account_json))
+
+        self._service = discovery.build(
+                SERVICE_NAME,
+                API_VERSION,
+                discoveryServiceUrl=discovery_url(api_key),
+                credentials=credentials)
+
+        # Lookup or create the device registry. Here we bind the registry to
```
Comment (jonparrott) [inline iot/api-client/manager/cloudiot_device_manager_example.py]: This seems to constitute a separate function, `create_or_get_registry`.

File: iot/api-client/manager/cloudiot_device_manager_example.py [hunk 6]
```diff
@@ -0,0 +1,330 @@
...
+DISCOVERY_API = 'https://cloudiot.googleapis.com/$discovery/rest'
+SERVICE_NAME = 'cloudiot'
+
+
+def discovery_url(api_key):
+    """Construct the discovery url for the given api key."""
+    return '{}?version={}&key={}'.format(DISCOVERY_API, API_VERSION, api_key)
+
+
+class DeviceRegistry(object):
```
Comment (jonparrott) [inline iot/api-client/manager/cloudiot_device_manager_example.py]: This class reads like something from a client library, not like something from a code sample. I would consider how you could reduce this class to stateless functions and how you could consider making the code flow through logical steps such as create registry, register a device, etc.

File: iot/api-client/manager/cloudiot_device_manager_example.py [hunk 7]
```diff
@@ -0,0 +1,330 @@
...
+                }
+            }]
+        }
+        return self._create_device(device_id, device_template)
+
+    def create_device_with_no_auth(self, device_id):
+        """Create a new device with no authentication."""
+        return self._create_device(device_id, {})
+
+    def patch_es256_for_auth(self, device_id, public_key_file):
```
Comment (jonparrott) [inline iot/api-client/manager/cloudiot_device_manager_example.py]: Recommendation: name this after the *intent*, not the implementation, so maybe `add_es256_key_to_device`.

File: iot/api-client/generate_keys.sh (hunk 2 above)
Comment (gguuss) [inline iot/api-client/generate_keys.sh]: Works from Cygwin / Bash for Windows AFAIK.  For a native solution, I have used PuTTyGen in the past.

File: iot/api-client/manager/cloudiot_device_manager_example.py (hunk 6 above)
Comment (gguuss) [inline iot/api-client/manager/cloudiot_device_manager_example.py]: Current plan is to change this to a CLI before we exit EAP, I consider it an exercise outside of the scope of the initial commit.

File: iot/api-client/README.md (hunk 1 above)
Comment (gguuss) [inline iot/api-client/README.md]: I am reusing the parent README in Node / Java / Python because there's some tricky steps that are unrelated to the language-specific steps.

**Comment by jmdobry** [comment]
FYI: The TODO-list we came up with for Node.js after the initial merge: https://github.com/GoogleCloudPlatform/nodejs-docs-samples/pull/394#issuecomment-303513026

File: iot/api-client/manager/cloudiot_device_manager_example.py (hunk 4 above)
Comment (gguuss) [inline iot/api-client/manager/cloudiot_device_manager_example.py]: Updated but now I'm seeing:

"AttributeError: 'Credentials' object has no attribute 'authorize'"

**Comment by jonparrott** [comment]
@dpebot will you merge when travis passes?
